src/unif.ml

```diff
@@ -102,28 +116,63 @@ let unify_modulo : term -> term -> bool = fun a b ->
   if !debug_equa then log "unif" "%a == %a" pp a pp b;
   let rec unify_modulo l =
     match l with
-    | []                   -> true
+    | []                      -> true
     | (a,b)::l when unify a b -> unify_modulo l
-    | (a,b)::l             ->
-        let (a,sa) = whnf_stk a [] in
-        let (b,sb) = whnf_stk b [] in
-        let rec sync acc la lb =
-          match (la, lb) with
-          | ([]   , []   ) -> (a, b, acc)
-          | (a::la, b::lb) -> sync ((!a,!b)::acc) la lb
-          | (la   , []   ) -> (to_term a (List.rev la), b, acc)
-          | ([]   , lb   ) -> (a, to_term b (List.rev lb), acc)
-        in
-        let (a,b,l) = sync l (List.rev sa) (List.rev sb) in
-        match (a, b) with
-        | (a          , b          ) when unify a b -> unify_modulo l
-        | (Abst(aa,ba), Abst(ab,bb)) ->
-            let (_,ba,bb) = Bindlib.unbind2 mkfree ba bb in
-            unify_modulo ((aa,ab)::(ba,bb)::l)
-        | (Prod(aa,ba), Prod(ab,bb)) ->
-            let (_,ba,bb) = Bindlib.unbind2 mkfree ba bb in
-            unify_modulo ((aa,ab)::(ba,bb)::l)
-        | (a          , b          ) ->
+    | (a,b)::l                ->
+    let a = whnf a in
+    let b = whnf b in
+    let rec get_args args a b =
+      match (unfold a, unfold b) with
+      | (Appl(t1,u1), Appl(t2,u2)) -> get_args ((u1,u2)::args) t1 t2
+      | (Meta(u1,e1), Meta(u2,e2)) when u1 == u2 -> (a,b,args)
+      | (Meta(m,e)  , b          ) when instantiate m e b -> get_args args a b
+      | (a          , Meta(m,e)  ) when instantiate m e a -> get_args args a b
+      | (_          , Meta(_,_)  ) -> assert false
+      | (_          , _          ) -> (a,b,args)
+    in
+    let unif =
```

Comment:
Why not always using unify_modulo? Comment required.

Reply:
Using `unify_modulo` is necessary when not in constraint mode. However, it should not be used in constraint mode, as otherwise we somehow bypass the injectivity check for constraints.